dll/win32/shlwapi/autocomp.cpp

```diff
@@ -0,0 +1,616 @@
+    void DoURLMRU();
+
+    /* IEnumString interface */
+    STDMETHODIMP Next(ULONG celt, LPOLESTR *rgelt, ULONG *pceltFetched);
+    STDMETHODIMP Skip(ULONG celt);
+    STDMETHODIMP Reset();
+    STDMETHODIMP Clone(IEnumString **ppenum);
+
+    BEGIN_COM_MAP(CAutoCompleteEnumString)
+        COM_INTERFACE_ENTRY_IID(IID_IEnumString, IEnumString)
+    END_COM_MAP()
+
+protected:
+    INT m_iItem; // The position for Next and Skip
+    HWND m_hwndEdit; // The EDIT control
+    DWORD m_dwSHACF; // The SHACF_* flags
+    CSimpleArray<CStringW> m_items; // The items we got
+    DWORD m_dwOldTicks; // For checking time span
+};
+
+void CAutoCompleteEnumString::Initialize(IAutoComplete2 *pAC2, DWORD dwSHACF, HWND hwndEdit)
+{
+    m_hwndEdit = hwndEdit;
+    m_dwSHACF = dwSHACF;
+    Reset(); // Populate the items
+}
+
+STDMETHODIMP
+CAutoCompleteEnumString::Next(ULONG celt, LPOLESTR *rgelt, ULONG *pceltFetched)
+{
+    if (!rgelt || !pceltFetched)
+    {
+        ERR("E_POINTER\n");
+        return E_POINTER;
+    }
+
+    // Initialize
+    *pceltFetched = 0;
+    *rgelt = NULL;
+
+    if (m_iItem >= m_items.GetSize())
+        return S_FALSE; // No more elements
+
+    // Get the elements
+    ULONG ielt;
+    for (ielt = 0; ielt < celt && m_iItem < m_items.GetSize(); ++ielt, ++m_iItem)
+    {
+        size_t cb = (wcslen(m_items[m_iItem]) + 1) * sizeof(WCHAR);
+        rgelt[ielt] = (LPWSTR)CoTaskMemAlloc(cb);
+        if (!rgelt[ielt]) // Failed
+        {
+            ERR("Out of memory\n");
+            // Clean up
+            while (ielt-- > 0)
+            {
+                CoTaskMemFree(rgelt[ielt]);
+                rgelt[ielt] = NULL;
+            }
+            return S_FALSE; // Failure
+        }
+        CopyMemory(rgelt[ielt], (LPCWSTR)m_items[m_iItem], cb);
+    }
+    *pceltFetched = ielt; // The number of elements we got
+    return (ielt == celt) ? S_OK : S_FALSE; // Success or failure
+}
+
+STDMETHODIMP CAutoCompleteEnumString::Skip(ULONG celt)
+{
+    if (m_items.GetSize() < INT(m_iItem + celt))
+        return S_FALSE; // Out of bound
+
+    m_iItem += celt;
+    return S_OK; // Success
+}
+
+// https://stackoverflow.com/questions/3098696/get-information-about-disk-drives-result-on-windows7-32-bit-system/3100268#3100268
+static BOOL
+GetDriveCharacteristics(HANDLE hDevice, ULONG *pCharacteristics)
+{
+    NTSTATUS Status;
+    IO_STATUS_BLOCK IoStatusBlock;
+    FILE_FS_DEVICE_INFORMATION DeviceInfo;
+
+    Status = NtQueryVolumeInformationFile(hDevice, &IoStatusBlock,
+                                          &DeviceInfo, sizeof(DeviceInfo),
+                                          FileFsDeviceInformation);
+    if (Status == NO_ERROR)
+    {
+        *pCharacteristics = DeviceInfo.Characteristics;
+        return TRUE;
+    }
+    return FALSE;
+}
+
+static DWORD GetDriveFlags(LPCWSTR pszRoot)
+{
+    WCHAR szDevice[16];
+    HANDLE hDevice;
+    ULONG ret = 0;
+
+    lstrcpynW(szDevice, L"\\\\.\\", _countof(szDevice));
```

Comment:
I would prefer String[Cch|Cb] functions in new code.

Reply:
OK.